internal/goro/adaptive_pool.go

```diff
@@ -0,0 +1,135 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package goro
+
+import (
+	"math/rand"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common/clock"
+)
+
+type (
+	// AdaptivePool manages a pool of goroutines to handle small items of work.
+	// The size of the pool starts with minWorkers but can grow if needed, up to maxWorkers,
+	// and can shrink back down to minWorkers.
+	AdaptivePool struct {
```

Comment:
This is very similar to the dynamic_worker_pool_scheduler that I wrote.
I was also considering adding a timer before shrinking the pool.
I'd recommend merging these implementations.

Reply:
After reading the implementation a bit more I see there are some subtle differences. I don't know if it's worth trying to merge.